pkg/graveler/committed/skip_prefix_iterator.go

```diff
@@ -0,0 +1,168 @@
+	r  *Range
+	vr *graveler.ValueRecord
+}
+
+type ImportIterator interface {
+	IsCurrentRangeBoundedByPrefix() bool
+	IsCurrentPrefixIncludedInRange() bool
+}
+
+type SkipPrefixIterator struct {
+	currentPrefixIndex int
+	prefixes           []graveler.Prefix
+	rangeIterator      Iterator
+	currentRangeValue  rangeValue
+}
+
+func (ipi *SkipPrefixIterator) Value() (*graveler.ValueRecord, *Range) {
+	return ipi.currentRangeValue.vr, ipi.currentRangeValue.r
+}
+
+func (ipi *SkipPrefixIterator) updateValue() (*graveler.ValueRecord, *Range) {
+	vr, r := ipi.rangeIterator.Value()
+	ipi.currentRangeValue = rangeValue{
+		r,
+		vr,
+	}
+	return vr, r
+}
+
+func (ipi *SkipPrefixIterator) Err() error {
+	return ipi.rangeIterator.Err()
+}
+func (ipi *SkipPrefixIterator) Close() {
+	ipi.rangeIterator.Close()
+}
+func (ipi *SkipPrefixIterator) SeekGE(id graveler.Key) {
+	ipi.rangeIterator.SeekGE(id)
+}
+
+func (ipi *SkipPrefixIterator) Next() bool {
+	hasNext := ipi.rangeIterator.Next()
+	if !hasNext {
+		return false
+	}
+	vr, r := ipi.updateValue()
+	ipi.updatePrefix()
+
+	if vr == nil && r != nil { // head of range
+		for ipi.IsCurrentRangeBoundedByPrefix() {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.NextRange()
+			if !hasNext {
+				return false
+			}
+			ipi.updateValue()
+			ipi.updatePrefix()
+		}
+	} else {
+		prefixLen := len(ipi.prefixes)
+		for vr != nil && ipi.currentPrefixIndex < prefixLen && strings.HasPrefix(vr.Key.String(), string(ipi.prefixes[ipi.currentPrefixIndex])) {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.Next()
+			if !hasNext {
+				return false
+			}
+			vr, _ = ipi.updateValue()
+			ipi.updatePrefix()
+		}
+	}
+	return true
+}
+
+func (ipi *SkipPrefixIterator) NextRange() bool {
+	hasNext := ipi.rangeIterator.NextRange()
+	if !hasNext {
+		return false
+	}
+	ipi.updateValue()
+	ipi.updatePrefix()
+
+	for ipi.IsCurrentRangeBoundedByPrefix() {
+		if ipi.rangeIterator.Err() != nil {
+			return false
+		}
+		hasNext = ipi.rangeIterator.NextRange()
+		if !hasNext {
+			return false
+		}
+		ipi.updateValue()
+		ipi.currentPrefixIndex++
+	}
+	return true
+}
+
+func (ipi *SkipPrefixIterator) updatePrefix() {
+	if ipi.currentPrefixIndex >= len(ipi.prefixes) {
+		return
+	}
+	currMinKey := string(ipi.currentRangeValue.r.MinKey)
+	if ipi.currentRangeValue.vr != nil {
+		currMinKey = string(ipi.currentRangeValue.vr.Key)
+	}
+	// If the current prefix is smaller or isn't the prefix of the currentMinKey, get the next prefix.
+	// By the end of this loop, the examined prefix will either be the prefix of the currentMinKey, or
+	// lexicographically bigger than it.
+	for string(ipi.prefixes[ipi.currentPrefixIndex]) < currMinKey &&
+		!strings.HasPrefix(currMinKey, string(ipi.prefixes[ipi.currentPrefixIndex])) {
+		p := ipi.currentPrefixIndex + 1
+		switch {
+		case p >= len(ipi.prefixes): // No more comparable prefixes
+			return
```

Comment:
```suggestion
			ipi.currentPrefixIndex = p
			return
```
So next time function will return fast?